arcgis-ios-sdk-samples/Layers/List KML contents/ListKMLContentsSceneViewController.swift

```diff
@@ -0,0 +1,168 @@
+// Copyright 2018 Esri.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+class ListKMLContentsSceneViewController: UIViewController {
+    
+    @IBOutlet weak var sceneView: AGSSceneView!
+    
+    weak var kmlDataset: AGSKMLDataset?
+    weak var node: AGSKMLNode? {
+        didSet {
+            title = node?.name
+        }
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        guard let kmlDataset = kmlDataset else {
+            return
+        }
+        
+        // initialize scene with labeled imagery
+        let scene = AGSScene(basemapType: .imageryWithLabels)
+        
+        // create an elevation source and add it to the scene's surface
+        let elevationSourceURL = URL(string: "https://elevation3d.arcgis.com/arcgis/rest/services/WorldElevation3D/Terrain3D/ImageServer")!
+        let elevationSource = AGSArcGISTiledElevationSource(url: elevationSourceURL)
+        scene.baseSurface?.elevationSources.append(elevationSource)
+        
+        // assign the scene to the view
+        sceneView.scene = scene
+        
+        // create a layer to display the dataset
+        let kmlLayer = AGSKMLLayer(kmlDataset: kmlDataset)
+        
+        // add the kml layer to the map
+        scene.operationalLayers.add(kmlLayer)
+        
+        setSceneViewpoint()
+    }
+    
+    //MARK: - Viewpoint
+    
+    private func setSceneViewpoint(){
+        guard let node = node else {
+            return
+        }
+        if let nodeViewpoint = viewpoint(for: node),
+            !nodeViewpoint.targetGeometry.isEmpty {
+            // set the viewpoint for the node
+            sceneView.setViewpoint(nodeViewpoint)
+        }
+        else {
+            // this node has no viewpoint so hide the scene view, showing the info text
+            sceneView.isHidden = true
+        }
+    }
+    
+    /// Returns the elevation of the scene's surface corresponding to the point's x/y.
+    private func sceneSurfaceElevation(for point: AGSPoint) -> Double? {
+        guard let surface = sceneView.scene?.baseSurface else {
+            return nil
+        }
+        
+        var surfaceElevation: Double? = nil
+        let group = DispatchGroup()
+        group.enter()
+        // we want to return the elevation synchronously, so run the task in the background and wait
+        DispatchQueue.global(qos: .userInteractive).async {
```

Comment:
I use the group to wait for the inner closure to return, so using `sync` wouldn't work. The purpose of the outer `async` is to keep the thread from blocking since the inner closure is run on the same thread that deploys it.

Reply:
Oh! That makes perfect sense. Never mind.